### PR TITLE
Implement drop-at-a-distance semantics.

### LINKED ISSUE #690: Dropped node doesn't realise it is dropped
```
make cluster -j20 TMUX_LAYOUT=tiled NODES=3
pg_autoctl drop node --hostname localhost --pgport 5502 --pgdata monitor
```

Continues logging the following line every second:
```
16:35:59 29172 INFO  New state for node 1 "node1" (localhost:5501): primary ➜ primary                                                                                                                                                                                     
16:36:08 29172 INFO  New state for node 1 "node1" (localhost:5501): primary ➜ apply_settings                                                                                                                                                                              
16:36:08 29172 ERROR Monitor ERROR:  couldn't find node with nodeid 2                                                                                                                                                                                                     
16:36:08 29172 ERROR Failed to get node state for node 2 in group 0 of formation "default" with initial state "secondary", replication state "", and current lsn "0/4000110", see previous lines for details                                                              
16:36:08 29172 ERROR Failed to get the goal state from the monitor                                                                                                                                                                                                        
16:36:09 29172 INFO  New state for node 1 "node1" (localhost:5501): apply_settings ➜ apply_settings                                                                                                                                                                       
16:36:09 29172 INFO  New state for node 1 "node1" (localhost:5501): apply_settings ➜ primary                                                                                                                                                                              
16:36:09 29172 ERROR Monitor ERROR:  couldn't find node with nodeid 2                                                                                                                                                                                                     
16:36:09 29172 ERROR Failed to get node state for node 2 in group 0 of formation "default" with initial state "secondary", replication state "", and current lsn "0/4000110", see previous lines for details                                                              
16:36:09 29172 ERROR Failed to get the goal state from the monitor                                                                                                                                                                                                        
16:36:09 29172 INFO  New state for node 1 "node1" (localhost:5501): primary ➜ primary                                                                                                                                                                                     
16:36:09 29172 ERROR Monitor ERROR:  couldn't find node with nodeid 2                                                                                                                                                                                                     
16:36:09 29172 ERROR Failed to get node state for node 2 in group 0 of formation "default" with initial state "secondary", replication state "", and current lsn "0/40001C0", see previous lines for details                                                              
16:36:09 29172 ERROR Failed to get the goal state from the monitor                                                                                                                                                                                                        
16:36:10 29172 ERROR Monitor ERROR:  couldn't find node with nodeid 2                                                                                                                                                                                                     
16:36:10 29172 ERROR Failed to get node state for node 2 in group 0 of formation "default" with initial state "secondary", replication state "", and current lsn "0/40001C0", see previous lines for details                                                              
16:36:10 29172 ERROR Failed to get the goal state from the monitor                                                                                                                                                                                                        
16:36:11 29172 ERROR Monitor ERROR:  couldn't find node with nodeid 2                                                                                                                                                                                                     
16:36:11 29172 ERROR Failed to get node state for node 2 in group 0 of formation "default" with initial state "secondary", replication state "", and current lsn "0/40001C0", see previous lines for details                                                              
16:36:11 29172 ERROR Failed to get the goal state from the monitor                                                                                                                                                                                                        
16:36:12 29172 ERROR Monitor ERROR:  couldn't find node with nodeid 2                                                                                                                                                                                                     
16:36:12 29172 ERROR Failed to get node state for node 2 in group 0 of formation "default" with initial state "secondary", replication state "", and current lsn "0/40001C0", see previous lines for details                                                              
16:36:12 29172 ERROR Failed to get the goal state from the monitor                                                                                                                                                                                                        
```

Not sure if related, but valgrind shows the following:
```
==31720== Jump to the invalid address stated on the next line
==31720==    at 0x556: ???
==31720==    by 0x4005BF1: open_verify.constprop.7 (dl-load.c:1723)
==31720==    by 0x40063B7: open_path (dl-load.c:2051)
==31720==    by 0x4009338: _dl_map_object (dl-load.c:2291)
==31720==    by 0x400E4C1: openaux (dl-deps.c:63)
==31720==    by 0x401BE4A: _dl_catch_exception (dl-error-skeleton.c:196)
==31720==    by 0x400E83A: _dl_map_object_deps (dl-deps.c:249)
==31720==    by 0x4003EC5: dl_main (rtld.c:1726)
==31720==    by 0x401AF4F: _dl_sysdep_start (dl-sysdep.c:253)
==31720==    by 0x4002127: _dl_start_final (rtld.c:414)
==31720==    by 0x4002127: _dl_start (rtld.c:521)
==31720==    by 0x4001097: ??? (in /lib/x86_64-linux-gnu/ld-2.27.so)
==31720==    by 0x2: ???
==31720==  Address 0x556 is not stack'd, malloc'd or (recently) free'd
```

### PR DESCRIPTION
**Pull request by DimCitus**
This allows running `pg_autoctl drop node --name foo` from the monitor, or
even using the SQL API directly, and have the node realise it's been
dropped. And then stop.

The configuration file, state file, and PGDATA are not touched, for that the
command pg_autoctl drop node --pgdata ... --destroy can be used on the node
itself.

Fixes #690

### CODE REVIEW

**Review by JelteF — APPROVED** [review]
Issues found:
1. `pg_autoctl drop node --destroy` is indeed broken if done without first running `pg_autoctl drop node`
2. In a two node cluster `pg_autoctl drop node` results in no dropped node and two `single` nodes
3. We still need a way to remove a node from the monitor, even if the node is completely down. Maybe add a flag like `--force`.

**Review by JelteF — CHANGES_REQUESTED** [review]
(wrong button before)

File: src/bin/pg_autoctl/service_keeper.c [hunk 1]
```diff
@@ -254,13 +254,39 @@ keeper_node_active_loop(Keeper *keeper, pid_t start_pid)
...
+
 	log_debug("pg_autoctl service is starting");
 
 	/* setup our monitor client connection with our notification handler */
 	(void) monitor_setup_notifications(monitor,
 									   keeperState->current_group,
 									   keeperState->current_node_id);
 
+	/*
+	 * When pg_autoctl drop node is used from a distance, then this nodes
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/service_keeper.c]: Fixed.

File: src/monitor/pgautofailover.sql [hunk 2]
```diff
@@ -40,7 +40,8 @@ CREATE TYPE pgautofailover.replication_state
     'wait_maintenance',
     'report_lsn',
     'fast_forward',
-    'join_secondary'
+    'join_secondary',
+    'dropped'
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/monitor/pgautofailover.sql:44]: Added a migration, still some rounds to go before next review though...

**Comment by JelteF** [comment]
Problems  found:
1. wait_standby -> single transition is broken
2. wait_standby -> dropped transition is broken
afbeelding

**Comment by JelteF** [comment]
3. When running `drop` on a postgres server (not the monitor), nothing happens for 30 seconds.
4. sync_state and WAL LSN are empty in the log
afbeelding

File: src/bin/pg_autoctl/fsm.c [hunk 3]
```diff
@@ -406,9 +416,8 @@ keeper_fsm_step(Keeper *keeper)
 	 */
 	if (!keeper_update_pg_state(keeper))
 	{
-		log_error("Failed to update the keeper's state from the local "
-				  "PostgreSQL instance, see above for details.");
-		return false;
```
Comment (JelteF) [inline src/bin/pg_autoctl/fsm.c:411]: Why did you remove this `return false`?

Comment (DimCitus) [reply src/bin/pg_autoctl/fsm.c:411]: The `keeper_fsm_step` function is only used outside of the main loop, from a client side pg_autoctl process. In that context, we don't want to error out when Postgres is not running just because we want to e.g. drop the local node.

**Comment by DimCitus** [comment]
> * When running `drop` on a postgres server (not the monitor), nothing happens for 30 seconds.

This is fixed now.

> * sync_state and WAL LSN are empty in the log

Yes, because Postgres is not running, but it shouldn't prevent us from dropping the node. That's fixed now too.

**Review by JelteF — CHANGES_REQUESTED** [review]
I still get this log with empty values when I do (everything still works, but the log looks confusing):
1. stop node
2. drop node from monitor
3. `pg_autoctl run` or "drop node locally again" (`drop node --pgdata node2`)
```
10:34:52 29061 INFO  Calling node_active for node default/4/0 with current state: PostgreSQL is running is false, sync_state is "", latest WAL LSN is 0/0.
```

File: src/bin/pg_autoctl/keeper_pg_init.c [hunk 4]
```diff
@@ -127,23 +127,49 @@ keeper_pg_init_and_register(Keeper *keeper)
...
+	 * If we have a state file, we're either running the same command again
+	 * (such as pg_autoctl create postgres --run ...) or maybe the user has
+	 * changed their mind after having done a pg_autoctl drop node.
+	 */
 	if (file_exists(config->pathnames.state))
 	{
-		if (createAndRun)
+		bool dropped = false;
+
+		if (!keeper_node_has_been_dropped(keeper, &dropped) && dropped)
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/keeper_pg_init.c]: Good point, added error checking.

**Comment by DimCitus** [comment]
> ```
> 10:34:52 29061 INFO  Calling node_active for node default/4/0 with current state: PostgreSQL is running is false, sync_state is "", latest WAL LSN is 0/0.
> ```

Indeed, `keeper_fsm_step` didn't get the memo about this log output being a log_debug nowadays. Fixed, thanks!

File: src/monitor/pgautofailover--1.5--1.6.sql [hunk 5]
```diff
@@ -254,15 +254,33 @@ DROP FUNCTION pgautofailover.remove_node(int);
 
 CREATE FUNCTION pgautofailover.remove_node
  (
-   node_id bigint
+   node_id bigint,
+   force   bool default 'false'
```
Comment (JelteF) [inline src/monitor/pgautofailover--1.5--1.6.sql:258]: Migration mostly looks good, did you check `\dx+ pgautofailover` differences?

File: src/monitor/pgautofailover.sql (hunk 2 above)
Comment (JelteF) [inline src/monitor/pgautofailover.sql:44]: This new `dropped` one is not added in the migration.

File: src/bin/pg_autoctl/monitor.c [hunk 6]
```diff
@@ -1214,22 +1215,29 @@ monitor_set_formation_number_sync_standbys(Monitor *monitor, char *formation,
...
 	paramValues[0] = host;
 	paramValues[1] = intToString(port).strValue;
+	paramValues[2] = force ? "true" : "false";
 
 	if (!pgsql_execute_with_params(pgsql, sql,
 								   paramCount, paramTypes, paramValues,
 								   &context, &parseSingleValueResult))
 	{
+		/* if we fail to find the node we want to remove, we're good */
+		if (strcmp(context.sqlstate, STR_ERRCODE_UNDEFINED_OBJECT) == 0)
```
Comment (JelteF) [inline src/bin/pg_autoctl/monitor.c:1236]: Could this specific error code happen for other reasons? e.g. the function was removed during an upgrade.

File: src/bin/pg_autoctl/keeper.c [hunk 7]
```diff
@@ -1115,6 +1120,259 @@ keeper_maintain_replication_slots(Keeper *keeper)
...
+							   assignedState);
+}
+
+
+/*
+ * keeper_node_has_been_dropped checks if the local node is being dropped or
+ * has been dropped already from the monitor.
+ */
+bool
+keeper_node_has_been_dropped(Keeper *keeper, bool *dropped)
```
Comment (JelteF) [inline src/bin/pg_autoctl/keeper.c:1252]: This function does a bit more than the comment and name suggest. If the node is currently in the process of being dropped (assigned drop, but currently some other state), this process is finished by this function. Let's at least change the comment and probably also the name.

**Review by JelteF — CHANGES_REQUESTED** [review]
These two problems are still present
1. wait_standby -> single transition is broken (triggered by dropping primary in two node cluster, when other node is wait_standby)
2. wait_standby -> dropped transition is broken
afbeelding

I think it's fine to not fix issue 1, since it seems very rare. But issue 2 should be fixed, since it should be possible to drop a node that's currently in wait_standby mode.

File: src/monitor/pgautofailover.sql (hunk 2 above)
Comment (DimCitus) [reply src/monitor/pgautofailover.sql:44]: Good catch, fixed now!

File: src/bin/pg_autoctl/keeper.c (hunk 7 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/keeper.c]: Good point, edited the comments and renamed the function `keeper_ensure_node_has_been_dropped`.

File: src/monitor/pgautofailover--1.5--1.6.sql (hunk 5 above)
Comment (DimCitus) [reply src/monitor/pgautofailover--1.5--1.6.sql:306]: Yes, no diffs.

File: src/bin/pg_autoctl/cli_drop_node.c [hunk 8]
```diff
@@ -0,0 +1,848 @@
...
+	}
+
+	/*
+	 * Now, when the pg_autoctl keeper service is still running, wait until
+	 * it has reached the DROPPED/DROPPED state on-disk and then exited.
+	 */
+	pid_t pid = 0;
+
+	if (!wait_for_pid_to_exit(config->pathnames.pid, 30, &pid))
+	{
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/cli_drop_node.c:582]: Added.

File: src/bin/pg_autoctl/monitor.c (hunk 6 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/monitor.c:1236]: I believe that would raise a `undefined_function` error (see https://www.postgresql.org/docs/current/errcodes-appendix.html) rather than a very generic `undefined_object` one. I think we're good here.

**Comment by DimCitus** [comment]
> 2. wait_standby -> dropped transition is broken

Fixed now.

**Comment by JelteF** [comment]
> Fixed now.
Dropping a node while creating it still gives me errors. Now this:

afbeelding

File: src/monitor/pgautofailover--1.5--1.6.sql [hunk 9]
```diff
@@ -4,8 +4,66 @@
...
+DROP FUNCTION pgautofailover.get_other_nodes(int);
+
+DROP FUNCTION pgautofailover.get_other_nodes
+              (integer,pgautofailover.replication_state);
+
+DROP FUNCTION pgautofailover.current_state(text);
+DROP FUNCTION pgautofailover.current_state(text,int);
+
+DROP TRIGGER disable_secondary_check ON pgautofailover.formation;
+DROP FUNCTION pgautofailover.update_secondary_check() CASCADE;
```
Comment (JelteF) [inline src/monitor/pgautofailover--1.5--1.6.sql:25]: I think using CASCADE is pretty scary in a migration. It also doesn't seem to be necessary.
```suggestion
DROP FUNCTION pgautofailover.update_secondary_check();
```

**Review by JelteF — CHANGES_REQUESTED** [review]

**Comment by DimCitus** [comment]
> Dropping a node while creating it still gives me errors. Now this:

You need to be quite fast to get that one. Then when trying I was even faster apparently and could drop the node in the step before that. Now fixed.

File: src/monitor/pgautofailover--1.5--1.6.sql (hunk 9 above)
Comment (DimCitus) [reply src/monitor/pgautofailover--1.5--1.6.sql]: Good catch. I needed that while developing the script, and then realised it wasn't a good idea, as you're pointing out, and made it so that it's not needed anymore. Then forgot to actually remove the keyword. Oops. Done now.

**Comment by JelteF** [comment]
If you're even faster you get this:

afbeelding

File: .travis.yml [hunk 10]
```diff
@@ -35,6 +35,11 @@ matrix:
     - env: PGVERSION=13 TEST=ssl
     - env: PGVERSION=14 TEST=ssl
     - env: LINTING=true
+  allow_failures:
+    - env: PGVERSION=14 TEST=multi
+    - env: PGVERSION=14 TEST=single
+    - env: PGVERSION=14 TEST=monitor
+    - env: PGVERSION=14 TEST=ssl
```
Comment (JelteF) [inline .travis.yml:42]: I think this can be removed now that your other PR is merged.

**Review by JelteF — APPROVED** [review]
Feel free to fix the last very quick drop issue I provided, but maybe it's not really necessary.

**Review by JelteF — CHANGES_REQUESTED** [review]
found another bug:
afbeelding

File: .travis.yml (hunk 10 above)
Comment (DimCitus) [reply .travis.yml:42]: I think we should keep Postgres 14 as allowed to fail builds/tests until it's released as 14.0. It's nice to follow along and fix when a beta introduces incompatibilities, though as we've seen here, as Travis installs Postgres from the community repositories, this may happen anytime in any PR we're working on... I feel like this should not block us.

**Comment by DimCitus** [comment]
> 1. 2 node cluster, wait until stable
> 2. Drop node2
> 3. Drop node1
> 4. recreate node1

Fixed now.

**Review by JelteF — APPROVED** [review]
LGTM :shipit: 🎉
